### DISCUSSION TITLE
Profiling event scheduler overheads

### DISCUSSION THREAD

**Discussion by quasiquote1** [sparcians/map#259] [Q&A] [answered]
Hi all, I'm a new user and starting to play around with the examples (CoreModel in particular).

One of the things I'm starting to investigate is the simulator speeds - the metric Kilo Ticks Per Second (KTPS).

Is there a way I can get more insights into the scheduler overhead - how long it takes to process an event?   Can any of these overheads be amortized in some way (e.g., tuning compiler options, structure layouts, etc)?

Thanks!

**Comment by ghost** [comment, marked as answer]
Good questions and somewhat relates to issue #193, but you're more curious as to the overhead from the Scheduler itself.

First, regarding #193, I started the effort (in a local branch) to enable the framework to track:
1. Number of calls to a specific event
2. Profiling time for each event (time spent in the handlers)

Unfortunately, I didn't get too far with it (higher priority items).  The design I chose will allow profiling on/off on the fly without causing any performance degradation when disabled.  All related to the handler proxy stuff.  But I digress...

If you want to checkout scheduler overhead using a profiler, the framework does build and run on MacOS (including the examples).  If you have access to MacOS with Xcode, you can use the Instruments tool to get a good idea on where the time is spent in the framework.  I can tell you that the scheduler itself is very small in overhead, with sparta::PayloadEvent taking a chunk of time (due to STL list, but I have a solution for this in branch [payload_event_speedup](https://github.com/sparcians/map/tree/knutel/payload_event_speedup) that I need to test rigorously).  

The `sparta_core_example` is slow mostly because of the choice I made to copy vector instruction payloads around instead of one simple grouping of instruction objects (see [this line](https://github.com/sparcians/map/blob/1559c0af389fa200a4191237671bb901610ffcae/sparta/example/CoreModel/src/Fetch.cpp#L91)).   I could also take advantage of `std::move` to remove the copy, but that requires the Port classes to accept rvalues -- something else on my wish/todo list.  😄 

If you want to convince yourself of the speed of the simulation infrastructure. the [skeleton pipeline](https://sparcians.github.io/map/skeleton_example.html) might be a good choice to play with:
```
cd release/examples/SkeletonPipeline
./sparta_skeleton -p top.producer0.params.max_ints_to_send 10000000
```
On my AMD TR4 box, this was the performance output (took a little over 1 sec to drive 10M requests):
```
Running Complete
  Simulation Performace      : wall(1.3249), system(0.0000), user(1.3300)
  Scheduler Tick Rate  (KTPS): 7518.8  (1k ticks per second)
  Scheduler Event Rate (KEPS): 22556.4 KEPS (1k events per second)
  Scheduler Events Fired: 30000001
Run Successful!
Saving reports...
Automatic Simulation Summary [0,10000002]
  scheduler
      ticks                               = 10000001
      picoseconds                         = 10000001
      seconds                             = 1e-05
      milliseconds                        = 0.01
      microseconds                        = 10
      nanoseconds                         = 10000
      host_machine_user_runtime_seconds   = 1.33
      host_machine_system_runtime_seconds = 0
      host_machine_wall_runtime_seconds   = 1.325
      host_user_time_count_ms             = 1330
      host_system_time_count_ms           = 0
      host_wall_time_count_ms             = 1325
```

Finally, the optimization flags I've given LLVM and gcc might not be the most optimal.  If you have suggestions, I'm open to input!  We've been playing around with LTO with some success...

**Comment by ghost** [comment]
One more thing about the core example, try this:
```
./sparta_core_example -p top.cpu.core0.fetch.params.fetch_max_ipc true -i 10M
```
This will force the fetch block to always push as many instructions as it can down the pipe (in the above case, up to 4/cycle).  On my machine, the simulator retired 10M instructions in 14 seconds at an ipc rate at 2.0.  Not too shabby for something that has room for tuning.  😉

**Reply by quasiquote1** [reply]
Thanks @knute-sifive for this information.   Due to machine differences I'm only able to achieve about 240 kilo ticks per second, making my runtime around 21 seconds for your above command-line.

I also enabled gprof profiling and saw that most of the time was spent in the constructor for ExampleInst:

  %   cumulative   self              self     total           
 time   seconds   seconds    calls   s/call   s/call  name    
 11.36      1.41     1.41 10000050     0.00     0.00  core_example::ExampleInst::ExampleInst(sparta::decode::DecoderBase const&, core_example::ExampleInst::TargetUnit, unsigned int, bool, sparta::Clock const*, core_example::ExampleInst::Status)
  6.10      2.16     0.76 110005150     0.00     0.00  sparta::Scheduler::scheduleEvent(sparta::Scheduleable*, unsigned long, unsigned int, bool)
  4.04      2.66     0.50 30000006     0.00     0.00  sparta::PhasedPayloadEvent<sparta::GlobalEventProxy>::allocateProxy_(sparta::GlobalEventProxy const&)
  3.88      3.14     0.48        1     0.48    11.98  sparta::Scheduler::run(unsigned long, bool, bool)

Question: Does Sparta use any non-standard memory allocator (like tcmalloc)?   If not - how difficult would it be to use it?

Since you are at SiFive perhaps you can offer some insights into the KTPS and sim times that you guys are used to seeing for a more detailed CPU pipeline.   Can you offer some KTPS ranges if I was to model a much more complex CPU (say 4-wide OoO with branch prediction, Icache, and a semi-realistic memory model)?

**Reply by ghost** [reply]
Curious on the type of machine you're using...

We have used `tcmalloc` with Sparta before with success, but decided to pull it with the addition of the `sparta::SpartaSharedPointer` coupled with its allocator.  I did see a small performance gain at the time.  We even have a memory logger that did work with it.  

But if I remember correctly, we ran into valgrind issues, which can be suppressed, but decided to focus on other things...   It's not hard to add if you want to play around with it.  

Without divulging too much internal information, our simulators (based on Sparta) are very complex and accurate; we currently run (on average in a CPU farm) about 100 KTPS.  

FYI, simulation performance is tied to some simple rules when building a model that we follow:

1. Only do work when you need to (no polling for work)
2. Do the work as quickly as possible (efficient programming)
3. Smart caching of previous work.  (example: cache decoded instructions and reuse)
4. Avoid using objects that use system memory allocations (`std::list`, `std::vector`, etc -- can be helped with tcmalloc)
5. Pass pointers to objects instead of value passing of objects between units
6. Make use of [Sparta's SharedPointer](https://sparcians.github.io/map/classsparta_1_1SpartaSharedPointer.html), which was built for perf modeling
